vowpalwabbit/core/src/vw.cc

```diff
@@ -751,6 +751,8 @@ void VW::setup_example(VW::workspace& all, VW::example* ae)
   // Set the interactions for this example to the global set.
   ae->interactions = &all.interactions;
   ae->extent_interactions = &all.extent_interactions;
+
+  if (all.calculate_feature_space_hash) { ae->calculate_feature_space_hash(); }
```

Comment:
If this is lazily evaluated then we don't need to put it in setup_example (which would be a very good thing)

Reply:
where should it go then?

Reply:
if it is to go in the las reduction then we won't benefit from the parser thread, is my only thought

Reply:
You could make `calculate_feature_space_hash()` return the value and cache it if you'd like to avoid rehashing

Reply:
I'm confused by your suggestion, it does get calculated and set in example_predict

Reply:
`calculate_feature_space_hash();` returns void and not the hash that was calculated. I just mean change `calculate_feature_space_hash()` to return the hash itself and if the hash was already calculated then just return the cached value.

Reply:
how about just renaming `calculate_feature_space_hash` to `set_feature_space_hash` and if it is already calculated then the call will do nothing

Reply:
You still need to know it has been set or not then to use it?

If `calculate_feature_space_hash()` returns the hash then you are guaranteed to get the hash whether it was pre-calculated or not

Reply:
fair I can see that, updated